syntax = "proto3";

package ommx.v1;

import "ommx/v1/constraint.proto";
import "ommx/v1/decision_variables.proto";
import "ommx/v1/function.proto";

message Instance {
  message Description {
    optional string name = 1;
    optional string description = 2;
    repeated string authors = 3;
    // The application or library name that created this message.
    optional string created_by = 4;
  }
  Description description = 1;

  Function objective = 2;
  repeated Constraint constraints = 3;

  // A map of decision variable IDs to its attributes.
  //
  // - This must constain every decision variables used in the objective and constraints.
  // - This can contains a decision variable that is not used in the objective or constraints.
  // - IDs are not required to be sequential.
  map<uint64, DecisionVariableAttribute> decision_variables = 4;
}
